### PR TITLE
feat: Add ONNX plugin and ML-based track classifier example

### PR DESCRIPTION
**Pull request by irinaene**
This PR implements an example of a "track classifier" that reproduces the predictions of a deep neural network trained on CKF reconstructed tracks with the TrackML detector at constant 2T magnetic field. The ML model predictions are performed using the inference capabilities of the ONNX Runtime framework (https://github.com/Microsoft/onnxruntime).

* The ML prediction step is provided by the `MLTrackClassifier` class, which demonstrates how to interface with the C++ API of ONNX Runtime release 1.4.0.
* `CKFPerformanceWriter` is modified to allow using the `MLTrackClassifier` predictions instead of the truth-based method for the duplication rate plots. This is off by default and can be turned on by providing a CLI argument (`use-ml-track-classification`) to the CKF example executable.
* A demo ML model (deep neural network trained on CKF reconstructed tracks with the TrackML detector at constant 2T magnetic field that can label a track as good or duplicate based on a set of trajectory state properties, currently number of hits, number of outliers, and chi2/dof) is used as the default and is provided in `Examples/Run/Reconstruction/MLAmbiguityResolutionDemo.onnx` .

cc @XiaocongAi

### CODE REVIEW

**Comment by paulgessinger** [comment]
I think we should discuss the way you conditionally add additional compilation unit and shared library to the example binary. 

Usually we try to restrict the ifs and elses based on options to try to not be too deep in the cmake include chain. We've had problems with weird interdependencies between options that we got this way. 

Would there be any alternative? Maybe factorizing the example so that they end up as similar executables, but share a (potentially significant) portion of their code through other mechanisms?

**Comment by XiaocongAi** [comment]
Looking at the code, it might be possible to get rid of those ifs for the ONNX plugin in the `Examples/Io/Performance/CMakeLists.txt` if we pass to the `CKFPerformanceWriter` a std::function instead of the explicit track classifier. But I'm not very sure whether we could avoid the ifs in the `Examples/Run/Reconstruction/CMakeLists.txt`.

File: Plugins/Onnx/src/MLTrackClassifier.cpp
Comment (XiaocongAi) [inline Plugins/Onnx/src/MLTrackClassifier.cpp:15]: I'm thinking if it's possible to have a base class for the ML interface if the constructor and the ` runONNXInference` etc. are the same for different ML implementations (we probably will have a different ML implementation in the future). Then the `MLTrackClassifier` could be inherited from this base interface and it only needs to implement the `predictTrackLabel` function.

Comment (irinaene) [reply Plugins/Onnx/src/MLTrackClassifier.cpp:15]: Yes, the constructor and `runONNXInference` function don't really care about the internal details of the ML model that's being used for inference, so it should be possible to separate this "general" functionality from the prediction function (that has to be somewhat specific to the type of ML model used).

**Review by msmk0 — CHANGES_REQUESTED** [review]
Hi @irinaene, thank you for opening the PR.

I have a general request regarding the integration into the examples. Since `Onnx` is an external dependency that is not usually available, it should not be required to run the general performance tools. I saw that you compile the code differently when `Onnx` is available. That means the code is behaving differently in difference circumstances. To avoid this, I would suggest to move the `Onnx` specific output into separate modules.

**Comment by irinaene** [comment]
Hi @msmk0 !
I've been testing some of the ideas suggested by @XiaocongAi in previous comments.
In particular, passing a `std::function` to `CKFPerformanceWriter` (instead of a class instance) seems to work well and it allows to remove any compilation dependence on the plugin.
The dependence on the `Onnx` plugin and its various options can be moved to all be inside of the executable `RecCKFTracks` -- but this does mean that the compilation of the executable will depend on whether the `Onnx` plugin is available or not.

Is this along the lines of what you had in mind?
I'm also open to suggestions if you have a better idea for how to deal with this `Onnx` dependence issue.

**Comment by msmk0** [comment]
The typical approach that we have used so far would be to not communicate via interfaces, but via data. In your case, that would mean to implement `OnnxClassifier` module/algorithm that reads track candidates from the even store and produces a new classifier collection, e.g. as `IndexMultimap<double>` that maps track candidate indices to your classifier value. Then you can either give this as an additional input to the existing performance writer (which does not need to be linked with Onnx or the plugin) or add a separate performance writer that only considers the classifiers.

For the executable, you can use a preprocessor flag and you only include the relevant header file and add the algorithm to the sequencer if given.

**Comment by msmk0** [comment]
@XiaocongAi @irinaene As discussed in private, this should be switched to the `std::function`-based approach as an intermediate step. Moving to the EDM-based can then happen in a separate PR. I will mark this PR as work-in-progress until then.

**Comment by msmk0** [comment]
@irinaene What is the status on this PR?

**Comment by irinaene** [comment]
@msmk0 I've added the relevant code updates. All the dependencies on the Onnx plugin should now be only in the `RecCKFTracks` executable.

**Comment by asalzburger** [comment]
Can this be rebased, please?

File: Examples/Io/Performance/ActsExamples/Io/Performance/CKFPerformanceWriter.cpp
Comment (msmk0) [inline Examples/Io/Performance/ActsExamples/Io/Performance/CKFPerformanceWriter.cpp:174]: Optimization: you only need to create this temporary structure once per `writeT` call instead of for each trajectory iteration.

File: Examples/Io/Performance/ActsExamples/Io/Performance/CKFPerformanceWriter.hpp
Comment (msmk0) [inline Examples/Io/Performance/ActsExamples/Io/Performance/CKFPerformanceWriter.hpp:63]: Unused and should be removed.

File: Examples/Io/Performance/ActsExamples/Io/Performance/CKFPerformanceWriter.hpp
Comment (msmk0) [inline Examples/Io/Performance/ActsExamples/Io/Performance/CKFPerformanceWriter.hpp:68]: I would suggest to simplify this further to 
```suggestion
    /// function to check if neural network predicted track label is duplicate
    std::function<bool(std::vector<float>&)> duplicatedPredictor = nullptr;
```
, i.e. to remove all classifier configuration from the writer and only store the final predictor function.

File: Examples/Run/Reconstruction/CMakeLists.txt
Comment (msmk0) [inline Examples/Run/Reconstruction/CMakeLists.txt:44]: This requires the Onnx plugin to run the CKF example, but we need to be able to run this without. My suggestion is to add a preprocessor flag such that the Onnx-related parts in the executable are only used if that flag is set. You can add the `ActsExampleCKFTracks` example unconditionally in the CMake configuration and only add the `ActsPluginOnnx` dependency if enabled.

File: Examples/Run/Reconstruction/RecCKFTracks.cpp
Comment (msmk0) [inline Examples/Run/Reconstruction/RecCKFTracks.cpp:215]: As discussed above: this should probably be enabled via a preprocessor flag.

File: Plugins/Onnx/include/Acts/Plugins/Onnx/MLTrackClassifier.hpp
Comment (msmk0) [inline Plugins/Onnx/include/Acts/Plugins/Onnx/MLTrackClassifier.hpp:24]:
```suggestion
  enum class TrackLabels { eGood, eDuplicate, eFake };
```
To match the [naming guidelines](https://acts.readthedocs.io/en/latest/codeguide.html#n-6-enum-values-use-ecamelcase) and to correctly namespace the labels.

File: Plugins/Onnx/include/Acts/Plugins/Onnx/MLTrackClassifier.hpp
Comment (msmk0) [inline Plugins/Onnx/include/Acts/Plugins/Onnx/MLTrackClassifier.hpp:35]:
```suggestion
                                double decisionThreshProb) const;
```
Nitpicking: scalar values should be handed over by-value.

File: Plugins/Onnx/include/Acts/Plugins/Onnx/MLTrackClassifier.hpp
Comment (msmk0) [inline Plugins/Onnx/include/Acts/Plugins/Onnx/MLTrackClassifier.hpp:46]:
```suggestion
                   double decisionThreshProb) const;
```
Nitpicking: scalar values should be handed over by-value.

File: Plugins/Onnx/include/Acts/Plugins/Onnx/OnnxRuntimeBase.hpp
Comment (msmk0) [inline Plugins/Onnx/include/Acts/Plugins/Onnx/OnnxRuntimeBase.hpp:37]: Where does the returned pointer point to? Are the values manually allocated by Onnx? Who is responsible for deallocating them? Does the returned values have to be modifiable or could this be `const float*` instead?

If I understand the implementation correctly, this does not have to be a `public` method and can be switched to `protected` to avoid having implementation details publi.

File: Plugins/Onnx/src/MLTrackClassifier.cpp
Comment (msmk0) [inline Plugins/Onnx/src/MLTrackClassifier.cpp:18]:
```suggestion
  if ((0 <= decisionThreshProb) and (decisionThreshProb <= 1)) {
```
Readability: this still matches the mathematical notation `0 < x < 1`.

**Review by msmk0 — CHANGES_REQUESTED** [review]
Hi @irinaene, thank you for updating this PR and reducing the Onnx dependency.

I have a few further suggestions to avoid breaking the examples for people without Onnx. I also have a question regarding the memory allocation of the classifier. At the moment this looks like a memory leak to me and I would like to clarify this before we merge this in.

**Comment by msmk0** [comment]
@irinaene Could you briefly mention what your status on this is?

**Comment by irinaene** [comment]
Hi @msmk0 ! I'm slowly working through the changes you suggested. The memory allocation might take a while to investigate since I'm not very familiar with memory allocation in C/C++ and it'll take some time to dig into the OnnxRuntime APIs and figure out what they're doing.

**Comment by msmk0** [comment]
> Hi @msmk0 ! I'm slowly working through the changes you suggested. The memory allocation might take a while to investigate since I'm not very familiar with memory allocation in C/C++ and it'll take some time to dig into the OnnxRuntime APIs and figure out what they're doing.

Thanks for letting us know. Freel free to contact me directly if you need any help with the allocation part.

**Comment by irinaene** [comment]
Hi @msmk0 ! I finished adding all the code changes. Re memory allocation, I looked through the Onnx API and my understanding is that Onnx takes care of deallocating the pointers when the destructor for the various Onnx objects is called (https://github.com/microsoft/onnxruntime/blob/1f69a58105f320e9b79008698b4275fa3f113f2d/include/onnxruntime/core/session/onnxruntime_cxx_api.h#L87), so this is not something we would have to do. In addition to the requested changes, I also changed the return of the `runOnnxInference` function to mimic the input part and keep all the Onnx specific details contained within the function.

File: Examples/Run/Reconstruction/Common/CMakeLists.txt
Comment (msmk0) [inline Examples/Run/Reconstruction/Common/CMakeLists.txt:35]: You can call `target_link_libraries` multiple times. A simple way would therefore be to only add a `if`-clause for the onnx runtime, i.e.
```suggestion
target_link_libraries(
  ActsExamplesRecTracksCommon
  PUBLIC
    ActsExamplesFramework
    ActsExamplesCommon
    ActsExamplesDigitization
    ActsExamplesTrackFitting
    ActsExamplesTrackFinding
    ActsExamplesMagneticField
    ActsExamplesTruthTracking
    ActsExamplesIoCsv
    ActsExamplesIoPerformance)

if(ACTS_BUILD_PLUGIN_ONNX)
  target_compile_definitions(
    ActsExamplesRecTracksCommon PUBLIC ACTS_PLUGIN_ONNX)
  target_link_libraries(
    ActsExamplesRecTracksCommon PUBLIC ActsPluginOnnx)
else()
```
where I also added the per-target compile definition.

File: Examples/Io/Performance/ActsExamples/Io/Performance/CKFPerformanceWriter.hpp
Comment (msmk0) [inline Examples/Io/Performance/ActsExamples/Io/Performance/CKFPerformanceWriter.hpp:61]: I do not think you need the `useMLTrackClassifier` flag here. The function reference below can be used for that. By default, the `std::function` is initialized w/ `nullptr` and does not reference a valid function. You can check in the algorithm implementation whether a predictor is set via
```cpp
if (m_cfg.duplicatedPredictor) {
  ...
}
```
directly. This also avoids the corner case, where the `useMLTrackClassifier` flag is set, but the function is not actually set.

**Review by msmk0 — APPROVED** [review]
@irinaene Thank you for the changes and my apologies for the long delay.

You addressed all my comments and I only have a few minor clarification comments. You can either address them directly, or we merge this PR in as-is and address them in a follow-up PR. I'll leave that up to you.

**Comment by irinaene** [comment]
Hi @msmk0 , thanks -- I added a commit for the minor comments.

It also seems like there is some issue with the `macos` CI job -- it failed very quickly at the `Install dependencies` step.

**Comment by paulgessinger** [comment]
That's a transient error I believe, where it fails to download the dependency tarball. Can you push an empty commit to retrigger it?

**Comment by paulgessinger** [comment]
It's not transient. I hope it'll be fixed with #603

**Review by msmk0 — APPROVED** [review]

**Comment by msmk0** [comment]
@irinaene Thanks for the fast fixes. Let's wait for #603 and then merge it.

**Pull request merged**
